pkg/route/validation/validation.go

```diff
@@ -342,6 +387,75 @@ func validateWildcardPolicy(host string, policy routev1.WildcardPolicyType, fldP
 	return nil
 }
 
+var (
+	notAllowedHTTPHeaders        = []string{"strict-transport-security", "proxy", "cookie", "set-cookie"}
+	notAllowedHTTPHeaderSet      = sets.NewString(notAllowedHTTPHeaders...)
+	notAllowedHTTPHeadersMessage = fmt.Sprintf("the following headers may not be modified using this API: %v", strings.Join(notAllowedHTTPHeaders, ", "))
+)
+
+// validateHeaders verifies that the given slice of request or response headers
+// is valid using the given regexp.
+func validateHeaders(fldPath *field.Path, headers []routev1.RouteHTTPHeader, nameRegexpForHeaderValue *regexp.Regexp, nameErrorMessage string, valueRegexpForHeaderValue *regexp.Regexp, valueErrorMessage string) field.ErrorList {
+	allErrs := field.ErrorList{}
+	headersMap := map[string]struct{}{}
+	for i, header := range headers {
+		idxPath := fldPath.Index(i)
+
+		// Each action must specify a unique header.
+		_, alreadyExists := headersMap[header.Name]
+		if alreadyExists {
+			err := field.Duplicate(idxPath.Child("name"), header.Name)
+			allErrs = append(allErrs, err)
+		}
+		headersMap[header.Name] = struct{}{}
+
+		switch {
+		case len(header.Name) == 0:
+			{
+				err := field.Required(idxPath.Child("name"), "")
+				allErrs = append(allErrs, err)
+			}
+		case !nameRegexpForHeaderValue.MatchString(header.Name):
+			{
+				err := field.Invalid(idxPath.Child("name"), header.Name, nameErrorMessage)
+				allErrs = append(allErrs, err)
+			}
+		case len(header.Name) > maxHeaderNameSize:
+			{
+				err := field.Invalid(idxPath.Child("name"), header.Name, fmt.Sprintf("name exceeds the maximum length, which is %d", maxHeaderNameSize))
+				allErrs = append(allErrs, err)
+			}
+		case notAllowedHTTPHeaderSet.Has(strings.ToLower(header.Name)):
+			{
+				err := field.Forbidden(idxPath.Child("name"), notAllowedHTTPHeadersMessage)
+				allErrs = append(allErrs, err)
+			}
+		}
+		if header.Action.Type != "" && (header.Action.Type != routev1.Set && header.Action.Type != routev1.Delete) {
```

Comment:
Why do you allow action to be empty?

Reply:
I posted https://github.com/miheer/library-go/pull/2 with a fix for this as well as a couple new test cases for empty action type and empty header name and some generate cleanups.

Reply:
This thread was resolved by cherry-picking https://github.com/openshift/library-go/pull/1449/commits/4900baef2e711657a46e41d7a7cba3a5c50e3c81.  Validation now forbids an empty action type: https://github.com/openshift/library-go/blob/4900baef2e711657a46e41d7a7cba3a5c50e3c81/pkg/route/validation/validation.go#L451-L454